### DISCUSSION TITLE
Sourcing dependencies from AWS, S3 buckets

### DISCUSSION THREAD

**Discussion by SphenicPaul** [renovatebot/renovate#26388] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.105.0

### If you're self-hosting Renovate, select which platform you are using.

Azure DevOps (dev.azure.com)

### What is your question?

I'm currently using AWS, S3 buckets as a store/repo of some built/published packages, which include:
- **Ansible 'Collections'** (that I don't want to publish to the public, Ansible Galaxy/repo, and I also don't (yet) have my own local, Ansible Automation Hub/Repo
- **AWS, Lambda Function, Packages/Zips** (that get pulled into Terraform provisioning of Lambda functions)

In both cases, although, I'm looking at the Ansible 'Collections' first, I want to be able to have Renovate source the available packages (and their versions) from the S3 bucket directory, given the relevant AWS token/secret IDs (presumably set as environment variables), and then be able to update these in the relevant files (`requirements.yml` for Ansible).

Note that I've some custom/bespoke logic in my Ansible builds to download S3-sourced dependencies so my `requirements.yml` file(s), contain the S3 path itself, like so...

```
---

collections:
- community.crypto

- name: s3://poc-some-s3-bucket/ansible_collections/some-package-0.1.6.tar.gz
  type: file

- name: s3://poc-some-s3-bucket/ansible_collections/some-other-package-0.1.0.tar.gz
  type: file
```

...and whilst, I believe I've got renovate to determine the files to update, using a `regexManager`, like so....

```
    {
      "fileMatch": [
        "^.*[Rr]equirements\\.yml",  // Ansible, Collection, 'Requirements.yml' file(s)
      ],
      "matchStrings": [

        // 'requirements.yml' usages - Examples:
        //   - name: s3://poc-some-s3-bucket/ansible_collections/some-package-0.1.3.tar.gz
        " {0,1}- {0,1}name: {0,1}(?<depDirectoryPath>.*)\\/(?<depName>.*)\\-(?<currentValue>.*?).tar.gz",

      ],
      "datasourceTemplate": "galaxy-collection", // File share (S3 location)
      "registryUrlTemplate": "s3://poc-some-s3-bucket/ansible_collections/",
      "repositoryCacheType": "s3://poc-some-s3-bucket/ansible_collections/",
    },
```
... and have tried to play around with `datasourceTemplate`, `registryUrlTemplate` and `repositoryCacheType` values, I'm not able to have renovate determine the list of packages, and would have thought that there must be some way of doing this, that I'm missing.

My fallback option is to additonally package these outputs within NuGet packages, then also push to a separate NuGet gallery, and update dependencies/versions from there (using the NuGet sources I already have configured and working), but that seems like I'd be duplicating those files, just for this Renovate process.

Any thoughts/options welcome.

### Logs (if relevant)

n/a

**Comment by rarkins** [comment]
If there are any registry types which support the s3:// protocol for fetching dependencies then their corresponding "datasource" in Renovate will need to be enhanced to add support for that, with the `maven` datasource being an example.

If you're hoping someone else to implement this, you'll need to provide a public reproduction (which presumably means a public S3 bucket). Otherwise if you're interested to submit a PR you can go right ahead.
